### PR TITLE
feat(igx-autocomplete): add igxAutocomplete directive #3585

### LINKED ISSUE #3585: igx-autocomplete
### Description  
Create igx-autocomplete directive.
[Specification](https://github.com/IgniteUI/igniteui-angular/wiki/igxAutocomplete-Specification#Acceptance-criteria).

### PR DESCRIPTION
**Pull request by Lipata**
Closes #3585
[Specification](https://github.com/IgniteUI/igniteui-angular/wiki/igxAutocomplete-Specification#Acceptance-criteria)
[Samples](https://github.com/IgniteUI/igniteui-angular-samples/issues/922)
[Topic](https://github.com/IgniteUI/igniteui-docfx/issues/991)

### Additional information:
 - [ ] Bug fix
 - [x] New functionality
 - [x] Documentation
 - [x] Demos
 - [x] CI/CD

### Checklist:
 - [x] All relevant tags have been applied to this PR
 - [x] All functionalities has been implemented 
 - [x] This PR includes unit tests covering all the new code above
 - [x] This PR includes API docs for newly added methods/properties
 - [x] This PR includes `feature/README.MD` updates for the feature docs
 - [x] This PR includes general feature table updates in the root `README.MD`
 - [x] This PR includes `CHANGELOG.MD` updates for newly added functionality
 - [ ] This PR includes `ng update` migrations for the breaking changes
 - [ ] This PR includes behavioral changes and the feature specification has been updated with them

### CODE REVIEW

File: projects/igniteui-angular/src/lib/directives/autocomplete/autocomplete.directive.ts [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+     */
+    @HostListener('blur', ['$event'])
+    onBlur() {
+        this.close();
+    }
+
+    /**
+     * @hidden
+     */
+    handleFocus() {}
```
Comment (rkaraivanov) [inline projects/igniteui-angular/src/lib/directives/autocomplete/autocomplete.directive.ts]: Is this some override?

Comment (Lipata) [reply projects/igniteui-angular/src/lib/directives/autocomplete/autocomplete.directive.ts]: It was but now it is not needed. Thanks!

**Comment by Lipata** [comment]
There are two problems that still needs to be addressed:

- clicking in the input, while drop down is opened will close it, because of the closeOnOutsideClick is true and the input is outside the drop down. For this enhancement in the drop down closing event is needed and it should include browser event in the arguments.
- #3743 when this is fixed we should reposition the the drop down on each input, for cases when drop down is above the input. @rkaraivanov for the moment the autocomplete drop down default position is by below the input, do you think we need to change default behavior to automatically position drop down either above or below the input, depending on the available space?

**Comment by Lipata** [comment]
After #3789 is fixed then the change in the autocomplete is minor, so it will be applied together with that bug.

File: projects/igniteui-angular/src/lib/directives/autocomplete/autocomplete.directive.ts [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+     */
+    get parentElement(): HTMLElement {
+        return this.group ? this.group.element.nativeElement : this.nativeElement;
+    }
+
+    /**
+     * @hidden
+     * @internal
+     */
+    target;
```
[the remark this answers is not in the document]
Comment (damyanpetev) [reply projects/igniteui-angular/src/lib/directives/autocomplete/autocomplete.directive.ts]: Also should remain public and documented.

File: projects/igniteui-angular/src/lib/directives/autocomplete/autocomplete.directive.ts [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+        if (args.cancel) {
+            return;
+        }
+        this.model ? this.model.control.setValue(newValue) : this.nativeElement.value = newValue;
+        this.close();
+        this.nativeElement.click();
+    }
+
+    private highlightFirstItem() {
+        const focusedItem = this.dropdown.focusedItem;
```
Comment (damyanpetev) [inline projects/igniteui-angular/src/lib/directives/autocomplete/autocomplete.directive.ts]: `this.target.navigateFirst();` ?

File: projects/igniteui-angular/src/lib/directives/autocomplete/autocomplete.directive.ts [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+     * <!--set-->
+     * <input type="text" [igxAutocomplete]="townsPanel" [igxAutocompleteDisabled]="disabled"/>
+     * ```
+     * ```typescript
+     * // set
+     * public disabled = true;
+     * ```
+     */
+    @Input('igxAutocompleteDisabled')
+    get disabled() {
```
Comment (damyanpetev) [inline projects/igniteui-angular/src/lib/directives/autocomplete/autocomplete.directive.ts]: `: boolean`

File: projects/igniteui-angular/src/lib/directives/autocomplete/autocomplete.directive.ts [hunk 5]
```diff
@@ -0,0 +1,385 @@
...
+        return this._disabled;
+    }
+    set disabled(value) {
+        this._disabled = value;
+        this.close();
```
Comment (damyanpetev) [inline projects/igniteui-angular/src/lib/directives/autocomplete/autocomplete.directive.ts]: No need for the close and get/set. It's API-only, close can be called if needed when setting this.

File: src/app/autocomplete/autocomplete.sample.html [hunk 6]
```diff
@@ -0,0 +1,50 @@
+<div class="sample-wrapper">
+    <app-page-header title="Autocomplete"></app-page-header>
+    <section class="sample-content">
+        <article class="sample-column">
+            <form class="input-sample" action="/" method="POST" (ngSubmit)="onSubmit($event)">
+                <div class="box">
+                    <igx-input-group class="group">
+                        <igx-prefix igxRipple><igx-icon fontSet="material">place</igx-icon> </igx-prefix>
```
Comment (damyanpetev) [inline src/app/autocomplete/autocomplete.sample.html]: Drop those ripples

File: projects/igniteui-angular/src/lib/directives/autocomplete/autocomplete.directive.ts [hunk 7]
```diff
@@ -0,0 +1,370 @@
...
+        value.cancel = true; // Disable selection in the drop down, because in autocomplete we do not save selection.
+        const newValue = value.newSelection.value;
+        const args: IAutocompleteItemSelectionEventArgs = { value: newValue, cancel: false };
+        this.onItemSelected.emit(args);
+        if (args.cancel) {
+            return;
+        }
+        this.model ? this.model.control.setValue(newValue) : this.nativeElement.value = newValue;
+        this.close();
+        this.nativeElement.click();
```
Comment (damyanpetev) [inline projects/igniteui-angular/src/lib/directives/autocomplete/autocomplete.directive.ts]: You mean `focus()`?

File: CHANGELOG.md [hunk 8]
```diff
@@ -2,10 +2,19 @@
 
 All notable changes for each version of this project will be documented in this file.
 ## 7.2.0
-
 - **New component** `IgxHierarchicalGrid`:
     - Provides the ability to represent and manipulate hierarchical data in which each level has a different schema. Each level is represented by a component derived from **igx-grid** and supports most of its functionality. Read up more information about the IgxHierarchicalGrid in the official [documentation](https://www.infragistics.com/products/ignite-ui-angular/angular/components/hierarchicalgrid.html) or the [ReadMe](https://github.com/IgniteUI/igniteui-angular/tree/master/projects/igniteui-angular/src/lib/grids/hierarchical-grid/README.md)
 
+- **New component** `igxAutocomplete` - new directive that provides a way to enhance a text input by showing a panel of suggested options, provided by the developer.
```
Comment (damyanpetev) [inline CHANGELOG.md]: Not a big deal, but it's a new directive

File: src/app/autocomplete/autocomplete.sample.ts [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+                closeAnimation: null,
+                openAnimation: null,
+                verticalDirection: VerticalAlignment.Top,
+                verticalStartPoint: VerticalAlignment.Top
+            })
+        };
+        this.enabled = false;
+    }
+
+    onSubmit() {}
```
Comment (damyanpetev) [inline src/app/autocomplete/autocomplete.sample.ts]: You could do something with the submit, like display the form values.

File: projects/igniteui-angular/src/lib/directives/autocomplete/autocomplete.directive.ts [hunk 10]
```diff
@@ -0,0 +1,370 @@
...
+
+    /**
+     * Emitted after item from the drop down is selected
+     *
+     * ```html
+     * <input igxInput [igxAutocomplete]="townsPanel" (onItemSelected)='itemSelected()' />
+     * ```
+     */
+    @Output()
+    onItemSelected = new EventEmitter<any>();
```
Comment (rkaraivanov) [inline projects/igniteui-angular/src/lib/directives/autocomplete/autocomplete.directive.ts]: `new EventEmitter<IAutocompleteItemSelectionEventArgs>();`

File: projects/igniteui-angular/src/lib/directives/autocomplete/autocomplete.directive.ts [hunk 11]
```diff
@@ -0,0 +1,370 @@
...
+     * ```
+     */
+    @Input('igxAutocompleteSettings')
+    autocompleteSettings: AutocompleteOverlaySettings;
+
+    /**
+     * Emitted after item from the drop down is selected
+     *
+     * ```html
+     * <input igxInput [igxAutocomplete]="townsPanel" (onItemSelected)='itemSelected()' />
```
Comment (rkaraivanov) [inline projects/igniteui-angular/src/lib/directives/autocomplete/autocomplete.directive.ts]: Shouldn't the handler take an `$event`?

**Pull request merged**